### PR TITLE
Popover: update `@floating-ui` to latest version, remove custom fix for iframe positioning and scaling

### PR DESCRIPTION
**Pull request by ciampo**
## What?

With https://github.com/floating-ui/floating-ui/issues/2043 merged, this PR updates `@floating-ui` to its latest version and removes any custom fixes that around `Popover` positioning across different documents (ie. anchor inside iframe, reference in root document). (also see https://github.com/floating-ui/floating-ui/issues/1869 for a description of the original issue)

## Why?

The custom fix was supposed to be temporary, and hopefully `@floating-ui`'s latest updates will allow us to remove it.

## How?

 - Updated `@floating-ui`'s version
 - Removed custom fixes

### In the editor

Open the site editor, try editing a template (or the whole site). Select a block, make sure that the toolbar is anchored correctly to the selected block. Try scrolling, make sure that the toolbar scrolls as expected and then sticks at the edges of the viewport.

### On storybook

Open the `Popover`'s "With Slot outside IFrame" Storybook example, and make sure that the `Popover` follow its anchor's position correctly

### Testing Instructions for Keyboard

N/A

## Screenshots or screencast 

N/A

### CODE REVIEW

**Comment by gziolo** [comment]
It would be great to take it to the finish line as it's causing issues when handling WordPress core releases. @tellthemachines highlighted some aspects of it in the [Proposal: improve the editor tech workflow for major releases](https://make.wordpress.org/core/2023/08/01/proposal-improve-the-editor-tech-workflow-for-major-releases/) in the "Package ecosystem problems" section.

**Comment by ockham** [comment]
👋 @ciampo @jsnajdr @tyxla @noahtallen Any chance y'all could maybe prioritize this one? 😊 Seems like the pinned dependency is adding some extra friction to the WordPress release process 😅

**Comment by jsnajdr** [comment]
I'm going to have a look :+1:

**Comment by jsnajdr** [comment]
I just rebased the PR onto latest `trunk` (it was many months old). It now uses the latest `@floating-ui/react-dom` version 2.0.0, which is a major upgrade from the previous 1.x version.

There is a breaking change in version 2.0 where the `useFloating` hook no longer returns the `floating` and `reference` fields, they've been moved to `refs.setFloating` and `refs.setReference`. It was easy to update the 2-3 usages.

But currently, after the upgrade, the positioning doesn't work. A block toolbar is not aligned to the block, but to the entire document canvas. We need to figure out what's wrong. @mirka have you been looking at this issue, too?

Another thing that's very likely to be broken now is the "Zoom out mode" stuff that @ellatrix added in #47004. We'll need to test the behavior when the iframe is zoomed.

**Comment by jsnajdr** [comment]
> But currently, after the upgrade, the positioning doesn't work. A block toolbar is not aligned to the block, but to the entire document canvas.

This is fixed now. To take iframes into account, the "virtual elements" passed to Floating UI, i.e., the objects with custom `getBoundingClientRect` methods, also need a `contextElement` field. That field is accessed in order to look up its `ownerDocument`, its `window` and the `window`'s `frameElement`.

What remains is to fix the "zoom out mode". I suspect we're accounting for the iframe's scale twice now.

**Comment by jsnajdr** [comment]
I believe this upgrade is ready for review, testing, and merging.

The last issue I fixed today was the experimental "zoom out" mode in Site Editor. I needed to:
- remove the code that manually figures out the owner iframe scale and applies it to the reference element, in a custom `getBoundingClientRect` function. The iframe scale is applied automatically inside `floating-ui`.
- re-add the `scroll` handler to the owner iframe, to update positioning every time the iframe scrolls.
- patch the `getReferenceOwnerDocument` function to correctly find the `ownerDocument` when the `anchor` is a `VirtualElement` with a `.contextElement` field.

**How to test:**
Test all kinds of popovers inside Gutenberg. I specifically tested block toolbars on individual blocks, and the "in between" popovers both in regular Post Editor and also in Site Editor's zoom-out mode.

File: packages/components/src/popover/index.tsx [hunk 1]
```diff
@@ -406,17 +364,16 @@ const UnforwardedPopover = (
...
 			anchor,
 			anchorRef,
 			anchorRect,
 			getAnchorRect,
 			fallbackReferenceElement,
-			scale,
 		} );
 
-		referenceCallbackRef( resultingReferenceElement );
+		refs.setReference( resultingReferenceElement );
```
Comment (tyxla) [inline packages/components/src/popover/index.tsx:377]: This appears to be causing an error in the site editor:

Error message

To reproduce:

* Open the site editor
* Roll over some blocks
* You should see the error.

Funny enough, this is also caught by a couple of TS errors (at least in my local environment):

ts-error

**Review by tyxla — DISMISSED** [review]

File: packages/components/src/popover/index.tsx (hunk 1 above)
Comment (jsnajdr) [reply packages/components/src/popover/index.tsx:377]: That's very strange, are you sure you ran `npm install` after checking out this branch? The `refs.setReference` API is new in `@floating-ui/react-dom@2.0.0`. Seems like you don't have it installed, and neither the new types.

Comment (tyxla) [reply packages/components/src/popover/index.tsx:377]: Funny enough I did run `npm install`, but now after running it again, it worked well. The error is also gone. Disregard. 👍

**Review by tyxla — COMMENTED** [review]
Doing some testing, I spotted some clunkiness that I see only with this branch. When opening the post editor more menu in a small window, and scrolling aggressively to the top, you can see some of that weird jumping around. Previously, that wasn't there.

This branch:

https://github.com/WordPress/gutenberg/assets/8436925/52bb9f7a-1015-4f15-bf55-fb18712d1a0e

`trunk`:


https://github.com/WordPress/gutenberg/assets/8436925/14d7aa1d-edf0-48cf-b8c1-2cf2218a8305

Is this something we should be worried about?

**Comment by jsnajdr** [comment]
> scrolling aggressively to the top, you can see some of that weird jumping around.

That's indeed an interesting observation, I'll try to figure out why the jumping is happening. Maybe it has something to do with the `scroll` listener we talked about yesterday. Looking at Floating UI changelog, I noticed some changes about usage of Intersection Observers etc. Maybe there are _two_ scroll-detecting mechanisms in action, both trying to handle the same scrolling.

**Comment by ciampo** [comment]
Not sure if it helps, but I wonder if these lines could be related:

https://github.com/WordPress/gutenberg/blob/f42b785b44faea34eae2dfe5d4f210e9e19d599e/packages/components/src/popover/index.tsx#L331-L334

Notice, in particular, the `animationFrame` option. I originally had to add it to get smoother updates, but I wonder if it may be in conflict with something else now?

Related docs:
- https://floating-ui.com/docs/react#anchoring
- https://floating-ui.com/docs/autoUpdate#animationframe

**Comment by jsnajdr** [comment]
> Notice, in particular, the `animationFrame` option.

There are some news about `animationFrame` in the Floating UI changelog: https://github.com/floating-ui/floating-ui/releases/tag/%40floating-ui%2Fdom%401.4.0 It says that there is a new `layoutShift` option for `autoUpdate` that supersedes `animationFrame` and uses `IntersectionObserver` instead. And after removing `animationFrame`, the jumping that @tyxla reported becomes indeed much calmer.

`animationFrame: true` is now recommended only if you want to recompute the position in reaction to an ongoing animation of the `transform` CSS property.

By the way, maybe the Post Editor More menu could benefit from using the [`overscroll-behavior` CSS property](https://developer.mozilla.org/en-US/docs/Web/CSS/overscroll-behavior) which would disable the "rubber band" overscroll of the menu element.

**Comment by jsnajdr** [comment]
After spending some more time with the `animationFrame` option and the menu scrolling behavior, and testing various scenarios, I came up with the final configuration for `autoUpdate`:
```js
autoUpdate( r, f, u, {
  layoutShift: false,
  animationFrame: true,
} );
```
I'm explicitly disabling the new `layoutShift` update checking, which uses `IntersectionObserver` on the reference element, and using only `animationFrame`.

It seems that it was this combination of two update checks that caused the menu jumping. Using only one check fixes that.

Using `layoutShift: true, animationFrame: false`, i.e., the default configuration, also works fairly well, but it's not the best. The `IntersectionObserver` fires its updates with certain time lag, and the popover repositioning also has a visible time lag when scrolling. For example, the block toolbar doesn't feel as it was "glued" to the block, but it's rather "flying behind" it.

File: packages/components/src/popover/index.tsx [hunk 2]
```diff
@@ -456,23 +408,17 @@ const UnforwardedPopover = (
...
-		const updateFrameOffset = () => {
-			frameOffsetRef.current = getFrameOffset( referenceOwnerDocument );
-			update();
-		};
-		defaultView.addEventListener( 'resize', updateFrameOffset );
-		scrollContainer?.addEventListener( 'scroll', updateFrameOffset );
-
-		updateFrameOffset();
+		defaultView.addEventListener( 'resize', update );
+		scrollContainer?.addEventListener( 'scroll', update );
```
Comment (jsnajdr) [inline packages/components/src/popover/index.tsx:412]: The reason why we need to attach our own listeners to the iframe's scroll container is described in https://github.com/floating-ui/floating-ui/issues/2518. Once that's fixed, Floating UI will attach them automatically for us, will watch for scroll position changes for us, and our workaround code can be removed.

**Comment by jsnajdr** [comment]
After addressing feedback, I once again believe this is ready for final review 🙂 @ciampo can you have an expert look at this? There surely are popover configurations I didn't test.

**Comment by ciampo** [comment]
Your findings about the `animationFrame` flag match my memories from when I worked on Popover last year - the `animationFrame` flag was the only way I could get popovers and dropdowns to render smoothly and follow the page while scrolling (even though they definitely come at a cost in terms of rendering perf)

I spent some time testing popovers across the editor and couldn't see any regression (although I admit that it's been 2 months since I last worked in the editor).

I say we merge this PR, and keep our eyes peeled for any report of regression, since we have a couple of months before the next WP release.

Feel free to approve this PR on my behalf (I can't, since I'm the original author)

**Review by jsnajdr — APPROVED** [review]

**Comment by ockham** [comment]
Thanks a lot for fixing this @ciampo @jsnajdr and @tyxla! 👏

**Comment by tellthemachines** [comment]
Thank you all so much for fixing this ❤️

**Pull request merged**